sdks/v4-sdk/src/utils/v4BaseActionsParser.ts

```diff
@@ -0,0 +1,209 @@
+                value: parseV4ExactIn(param),
+              }
+            case Subparser.V4SwapExactOutSingle:
+              return {
+                name: abiDef[j].name,
+                value: parseV4ExactOutSingle(param),
+              }
+            case Subparser.V4SwapExactOut:
+              return {
+                name: abiDef[j].name,
+                value: parseV4ExactOut(param),
+              }
+            case Subparser.PoolKey:
+              return {
+                name: abiDef[j].name,
+                value: parsePoolKey(param),
+              }
+            default:
+              return {
+                name: abiDef[j].name,
+                value: param,
+              }
+          }
+        })
+
+        return {
+          actionName: Actions[actionType],
+          actionType,
+          params,
+        }
+      }),
+    }
+  }
+
+  // parse command types from bytes string
+  private static getActions(actions: string): Actions[] {
+    const actionTypes = []
+
+    for (let i = 2; i < actions.length; i += 2) {
+      const byte = actions.substring(i, i + 2)
+      actionTypes.push(parseInt(byte, 16) as Actions)
+    }
+
+    return actionTypes
+  }
+}
+
+function parsePoolKey(data: string): PoolKey {
+  const [currency0, currency1, fee, tickSpacing, hooks] = data
+
+  return {
+    currency0,
+    currency1,
+    fee: parseInt(fee),
+    tickSpacing: parseInt(tickSpacing),
+    hooks,
+  }
+}
+
+function parsePathKey(data: string): PathKey {
+  const [intermediateCurrency, fee, tickSpacing, hooks, hookData] = data
+
+  return {
+    intermediateCurrency,
+    fee: parseInt(fee),
+    tickSpacing: parseInt(tickSpacing),
+    hooks,
+    hookData,
+  }
+}
+
+function parseV4ExactInSingle(data: any[]): SwapExactInSingle {
+  const [poolKey, zeroForOne, amountIn, amountOutMinimum, sqrtPriceLimitX96, hookData] = data
+  const [currency0, currency1, fee, tickSpacing, hooks] = poolKey
+  return {
+    poolKey: {
+      currency0,
+      currency1,
+      fee,
+      tickSpacing,
+      hooks,
+    },
+    zeroForOne,
+    amountIn,
+    amountOutMinimum,
+    sqrtPriceLimitX96,
+    hookData,
+  }
+}
+
+function parseV4ExactIn(data: any[]): SwapExactIn {
+  const [currencyIn, path, amountIn, amountOutMinimum] = data
+  const paths: readonly PathKey[] = path.map((pathKey: string) => parsePathKey(pathKey))
+
+  return {
+    path: paths,
+    currencyIn,
+    amountIn,
+    amountOutMinimum,
+  }
+}
+
+function parseV4ExactOutSingle(data: any[]): SwapExactOutSingle {
+  const [poolKey, zeroForOne, amountOut, amountInMaximum, sqrtPriceLimitX96, hookData] = data
+  const { currency0, currency1, fee, tickSpacing, hooks } = poolKey
+
+  return {
+    poolKey: {
+      currency0,
+      currency1,
+      fee,
+      tickSpacing,
+      hooks,
+    },
+    zeroForOne,
+    amountOut,
+    amountInMaximum,
+    sqrtPriceLimitX96,
+    hookData,
+  }
+}
+
+function parseV4ExactOut(data: any[]): SwapExactOut {
```

Comment:
did you check if this needs to be parsed in reverse order? i'd have to check in the contract but i know in v3 you have to reverse it

Reply:
hmm I'm not sure that matters in this context? I have a test that asserts the parsed path === the given path with the same order . Think the reversing is purely logical in how it's handled in the router but I'm just working with the calldata here right?

Reply:
https://github.com/Uniswap/sdks/blob/a501937d790f8ce963bd032712ae037b817db727/sdks/v4-sdk/src/utils/v4BaseActionsParser.test.ts#L268

Reply:
idk how we're going to display it to the person decoding but just meant - before we reversed v3, the parser interface used to say "USDC for WETH exact out" and it would actually be "WETH for USDC" so the trade was just wrong